src/modules/extra/m_argon2.cpp

```diff
@@ -0,0 +1,199 @@
+        }
+
+        // This isn't the raw version, but we don't have
+        // the facilities to juggle around the extra state required
+        // to do anything useful with them if we don't encode them.
+        // So we pretend this is the raw version, and instead make
+        // ToPrintable return its input.
+        auto result = std::string(encoded_data.data());
+        return result;
+    }
+
+    std::string ToPrintable(const std::string& raw) CXX11_OVERRIDE
+	{
+		return raw;
+	}
+
+    HashArgon2(Module* parent, const std::string& name, Argon2_type argon2Type)
+        : HashProvider(parent, name), argon2Type(argon2Type)
+    {
+        config = {};
+    }
+};
+
+class ModuleArgon2 : public Module
+{
+    HashArgon2 argon2i;
+    HashArgon2 argon2d;
+    HashArgon2 argon2id;
+    const std::array<HashArgon2*, 3> providers = { &argon2i, &argon2d, &argon2id };
+
+    static uint32_t SanitizeArgon2Version(const int version)
+    {
+        // Note, 10 is 0x10, and 13 is 0x13. Refering to it as
+        // dec 10 or 13 in the config file, for the name to
+        // match better.
+        if (version == 10) return ARGON2_VERSION_10;
+        if (version == 13) return ARGON2_VERSION_13;
+
+        ServerInstance->Logs->Log("MODULE", LOG_DEFAULT, "Unknown Argon2 version specified, assuming 13");
+        return ARGON2_VERSION_13;
+    }
+
+ public:
+    ModuleArgon2() :
+        argon2i(this, "argon2i", Argon2_i),
+        argon2d(this, "argon2d", Argon2_d),
+        argon2id(this, "argon2id", Argon2_id)
+    {
+    }
+
+    Version GetVersion() CXX11_OVERRIDE
+	{
+		return Version("Allows other modules to generate Argon2 hashes.", VF_VENDOR);
+	}
+
+    void ReadConfig(ConfigStatus& status) CXX11_OVERRIDE
+	{
+		auto tag = ServerInstance->Config->ConfValue("argon2");
+
+        ProviderConfigMap configs;
+        for (auto& provider : providers)
+        {
+            auto argonType = provider->name.substr(provider->name.find('/') + 1);
+            auto typeTag = ServerInstance->Config->ConfValue(argonType);
+            auto sourceTag = typeTag == ServerInstance->Config->EmptyTag ? tag : typeTag;
+            auto& config = configs[argonType];
+            ReadConfig(sourceTag, &config);
+        }
+
+        for (auto& config : configs)
+        {
+            auto provider = FindProvider(config.first);
+            provider->Configure(config.second);
+        }
+	}
+
+    HashArgon2* FindProvider(const std::string& name)
+    {
+        for (auto provider : providers)
+        {
+            auto hashName = provider->name.substr(provider->name.find('/') + 1);
+            if (name == hashName)
+            {
+                return provider;
+            }
+        }
+
+        // We know exactly what providers are loaded. We should always find what
+        // we're looking for.
+        throw ModuleException("Unable to find '" + name + "'. This is a bug.");
+    }
+
+    void ReadConfig(ConfigTag* tag, ProviderConfig* config)
```

Comment:
This config can be in the HashProvider then you can just call hashprovider.ReadConfig("argon2x") and get rid of the ProviderConfig class.

Reply:
I moved the logic and simplified it.
Didn't kill HashProvider though. I used it to hold the values from the argon2 block so it can be used as default values for the other config blocks (argon2i, argon2d, argon2id).